core/src/test/scala/unit/kafka/server/AbstractFetcherThreadTest.scala

```diff
@@ -346,6 +346,34 @@ class AbstractFetcherThreadTest {
     assertTrue(fetcher.fetchState(partition).get.isReadyForFetch)
   }
 
+  @Test
+  def testTruncateToHighWatermarkDuringRemovePartitions(): Unit = {
```

Comment:
Thanks for the test case. Maybe we can add a similar test case for the `truncateToEpochEndOffsets` path? I think `testLeaderEpochChangeDuringFetchEpochsFromLeader` has most of what we want, but this test case removes the partition and then adds it back.